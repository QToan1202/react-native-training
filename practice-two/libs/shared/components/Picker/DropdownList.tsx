import { useCallback } from 'react'
import { Stack, StackProps } from 'tamagui'
import { FlatList, ListRenderItem } from 'react-native'

import { IDropDownItem } from '@types'
import Button from '@components/Button'

export type DropDownListProps<T> = StackProps & {
  listData: T[]
  onSelect: (item: T) => void
}

const DropdownList = <T extends IDropDownItem>({
  listData,
  onSelect,
  ...rest
}: DropDownListProps<T>) => {
  const renderItem: ListRenderItem<T> = useCallback(
    ({ item }) => {
      const handleSetSelectItem = () => onSelect(item)

      return (
        <Button
          title={item.value}
          variant="tertiary"
          borderColor="$color.transparent"
          paddingHorizontal="$space.2.5"
          paddingVertical="$space.1.5"
          fontSize="$2"
          fontWeight="$4"
          color="$color.black"
          onPress={handleSetSelectItem}
        />
      )
    },
    [onSelect]
  )

  return (
    <Stack width="100%" alignItems="center" backgroundColor="$color.white" {...rest}>
      <FlatList
        data={listData}
        keyExtractor={({ id, value }: T): string => id || value}
        renderItem={renderItem}
      />
    </Stack>
  )
}

export default DropdownList
