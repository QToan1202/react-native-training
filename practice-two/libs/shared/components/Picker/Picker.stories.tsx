import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { Picker } from '../index'

export default {
  title: 'components/Picker',
  component: Picker,
} as ComponentMeta<typeof Picker>

const Template: ComponentStory<typeof Picker> = (args) => <Picker {...args} />

export const Default = Template.bind({})
Default.args = {
  listData: [
    {
      id: 'vn',
      name: 'vietnam',
      value: '+91',
    },
    {
      id: 'vn1',
      name: 'vietnam',
      value: '+92',
    },
    {
      id: 'vn2',
      name: 'vietnam',
      value: '+93',
    },
    {
      id: 'vn3',
      name: 'vietnam',
      value: '+94',
    },
    {
      id: 'vn4',
      name: 'vietnam',
      value: '+95',
    },
  ],
}
