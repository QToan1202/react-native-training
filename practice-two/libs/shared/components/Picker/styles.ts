import { StyleSheet } from 'react-native'

import { COLORS } from '@practice-two/shared'

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  spacing: {
    columnGap: 7,
  },
  selected: {
    width: 35,
    fontSize: 18,
    letterSpacing: -0.4,
  },
  list: {
    width: '100%',
    alignItems: 'center',
    backgroundColor: COLORS.WHITE,
  },
  listContent: {
    paddingHorizontal: 9,
    paddingVertical: 5,
    color: 'black',
    fontSize: 16,
  },
})

export default styles
