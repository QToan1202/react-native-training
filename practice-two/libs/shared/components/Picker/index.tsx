import { memo, useCallback, useMemo, useRef, useState } from 'react'
import isEqual from 'react-fast-compare'
import { Animated, ImageStyle, LayoutAnimation, StyleProp } from 'react-native'
import { XStack, XStackProps, YStack } from 'tamagui'

import Heading from '@components/Heading'
import { toggleAnim } from '@animations'
import { IDropDownItem } from '@types'

import DropdownList from './DropdownList'

export type PickerProps<T> = XStackProps & {
  listData: T[]
}

const Picker = <T extends IDropDownItem>({ listData, ...rest }: PickerProps<T>) => {
  const getFirstItemValue: string = useMemo(() => listData[0]?.value || ' ', [listData])
  const [selectedItem, setSelectedItem] = useState<string>(getFirstItemValue)
  const [isShow, setIsShow] = useState<boolean>(false)
  const animate: Animated.Value = useRef(new Animated.Value(0)).current

  const toggleDropdown = useCallback(() => {
    Animated.timing(animate, {
      duration: 300,
      toValue: isShow ? 0 : 1,
      useNativeDriver: true,
    }).start()
    LayoutAnimation.configureNext(toggleAnim)
    setIsShow((prevState) => !prevState)
  }, [animate, isShow])

  const arrowTransform = useRef<Animated.AnimatedInterpolation<string>>(
    animate.interpolate({
      inputRange: [0, 1],
      outputRange: ['0deg', '180deg'],
    })
  )

  const imageStyle: StyleProp<ImageStyle> = useMemo(
    () => ({
      transform: [
        {
          rotateZ: arrowTransform.current,
        },
      ],
    }),
    []
  )

  const handleSelectItem = useCallback(
    (item: T) => {
      setSelectedItem(item.value)
      toggleDropdown()
    },
    [toggleDropdown]
  )

  return (
    <YStack alignSelf="flex-start" overflow="hidden">
      <XStack alignItems="center" space="$space.2" onPress={toggleDropdown} {...rest}>
        <Heading content={selectedItem} fontSize="$3" letterSpacing="$4" width={35} />
        <Animated.Image source={require('@assets/arrow.png')} style={imageStyle} />
      </XStack>
      {isShow && <DropdownList listData={listData} onSelect={handleSelectItem} />}
    </YStack>
  )
}

export default memo(Picker, isEqual)
