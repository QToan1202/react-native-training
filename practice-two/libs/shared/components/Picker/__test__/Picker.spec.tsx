import { render, fireEvent, screen, act } from '@testing-library/react-native'

import Picker from '../index'

const listData = [
  {
    id: 'vn',
    name: 'vietnam',
    value: '+84',
  },
  {
    id: 'en',
    name: 'england',
    value: '+44',
  },
]

describe('Testing Picker component', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    render(<Picker listData={listData} testID="picker" />)
  })

  it('Render Picker component success', () => {
    const component = render(<Picker listData={listData} />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('First item of the list data will be selected on the init render', () => {
    const component = screen.getByText('+84')

    expect(component).toBeOnTheScreen()
  })

  it('Render component with empty list will show on the screen', () => {
    const { getByTestId } = render(<Picker listData={[]} testID="empty-picker" />)

    expect(getByTestId('empty-picker')).toBeOnTheScreen()
  })

  it('When tap on the picker show items as a dropdown component', async () => {
    const component = screen.getByTestId('picker')
    fireEvent.press(component)
    act(() => {
      jest.runAllTimers()
    })

    const firstDropdownItem = await screen.findAllByText('+84')
    const secondDropdownItem = await screen.findByText('+44')

    expect(firstDropdownItem.pop()).toBeOnTheScreen()
    expect(secondDropdownItem).toBeOnTheScreen()

    fireEvent.press(secondDropdownItem)
    expect(secondDropdownItem).not.toBeOnTheScreen()
  })
})
