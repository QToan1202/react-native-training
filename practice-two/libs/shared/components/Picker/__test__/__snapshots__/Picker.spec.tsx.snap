// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Picker component Render Picker component success 1`] = `
<View
  style={
    [
      {
        "alignSelf": "flex-start",
      },
      {
        "overflow": "hidden",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        },
        {
          "columnGap": 7,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_700Bold",
            "fontSize": 24,
          },
          false,
          {
            "fontSize": 18,
            "letterSpacing": -0.4,
            "width": 35,
          },
        ]
      }
    >
      +84
    </Text>
    <Image
      collapsable={false}
      source={
        {
          "testUri": "../../../assets/arrow.png",
        }
      }
      style={
        {
          "transform": [
            {
              "rotateZ": "0deg",
            },
          ],
        }
      }
    />
  </View>
</View>
`;
